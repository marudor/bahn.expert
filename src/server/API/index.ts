import { RegisterRoutes } from './routes';
import cors from '@koa/cors';
// import KoaRouter from '@koa/router';
import router from './validationOverwrites';

// const router = new KoaRouter();

router.use(
  cors({
    allowMethods: 'GET,POST',
    origin: '*',
    maxAge: 3600 * 24,
  }),
);
RegisterRoutes(router);
export default router;
