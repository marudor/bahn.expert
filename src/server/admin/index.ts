import Koa from 'koa';
import PromClient, { Counter, Histogram } from 'prom-client';
import type {
  AxiosInstance,
  AxiosResponse,
  InternalAxiosRequestConfig,
} from 'axios';
import type { Server } from 'node:http';

PromClient.register.clear();
PromClient.collectDefaultMetrics();

export const ApiRequestMetric = new Histogram({
  name: 'api_requests',
  help: 'api requests',
  labelNames: ['route', 'status'],
});

export const UpstreamApiRequestMetric = new Counter({
  name: 'upstream_api_requests',
  help: 'upstream api requests => bahn',
  labelNames: ['api'],
});

const UpstreamApiResponseMetric = new Counter({
  name: 'upstream_api_response',
  help: 'upstream api response => bahn',
  labelNames: ['api', 'code'],
});

function upstreamRequestApiCountInterceptor(
  apiName: string,
  req: InternalAxiosRequestConfig,
): InternalAxiosRequestConfig {
  UpstreamApiRequestMetric.inc({ api: apiName });
  return req;
}

function upstreamResponseApiCountInterceptor(
  apiName: string,
  res: AxiosResponse,
): AxiosResponse {
  UpstreamApiResponseMetric.inc({ api: apiName, code: res.status });
  return res;
}

export function axiosUpstreamInterceptor(
  axios: AxiosInstance,
  apiName: string,
): void {
  axios.interceptors.request.use(
    upstreamRequestApiCountInterceptor.bind(undefined, apiName),
  );
  axios.interceptors.response.use(
    upstreamResponseApiCountInterceptor.bind(undefined, apiName),
  );
}

export default (adminPort = 9000): Server => {
  const koa = new Koa();

  koa.use(async (ctx) => {
    try {
      switch (ctx.request.url) {
        case '/ping': {
          ctx.body = 'pong';
          break;
        }
        case '/metrics': {
          ctx.body = await PromClient.register.metrics();
          break;
        }
        default: {
          break;
        }
      }
    } catch {
      ctx.status = 500;
    }
  });

  return koa.listen(adminPort);
};
