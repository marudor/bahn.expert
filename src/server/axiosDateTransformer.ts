import type { AxiosResponseTransformer } from 'axios';

const isoDateRegex =
	/^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2}\.\d*)(?:Z|([+-])([\d:|]*))?$/;

export const axiosDateTransformer: AxiosResponseTransformer = (data) => {
	if (typeof data === 'string') {
		try {
			return JSON.parse(data, (_key, value) => {
				if (typeof value === 'string' && isoDateRegex.test(value)) {
					return new Date(value);
				}
				return value;
			});
		} catch {
			// Ignoring
		}
	}
	return data;
};
