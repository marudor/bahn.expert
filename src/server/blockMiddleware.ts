import type { Context } from 'koa';

const refererBlockList = process.env.REFERER_BLOCK_LIST?.split(',').map(
	(host) => new RegExp(`https?://${host}`),
);

export const blockMiddleware = (
	ctx: Context,
	next: () => Promise<any>,
): Promise<any> | undefined => {
	if (!refererBlockList) {
		return next();
	}
	const referer = ctx.request.headers.referer;
	if (referer && refererBlockList.some((r) => referer.match(r))) {
		ctx.status = 401;
		return;
	}
	return next();
};
