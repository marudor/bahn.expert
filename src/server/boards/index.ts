import { addMinutes, subMinutes } from 'date-fns';
import {
  getAbfahrten,
  sortAbfahrt,
  timeByReal,
  timeByScheduled,
} from '@/server/iris';
import { getGroups } from '@/server/StopPlace/search';
import Axios from 'axios';
import Qs from 'qs';
import type { Abfahrt, AbfahrtenResult } from '@/types/iris';

const privateAxios = Axios.create({
  baseURL: process.env.PRIVATE_API_URL,
  headers: {
    'x-api-key': process.env.PRIVATE_API_KEY || '',
  },
  paramsSerializer: {
    serialize: (params: any) =>
      Qs.stringify(params, {
        arrayFormat: 'repeat',
      }),
  },
});

export async function getPrivateAbfahrten(
  evaNumber: string,
  lookahead: number,
  lookbehind: number,
  startTime: Date = new Date(),
): Promise<AbfahrtenResult> {
  const stopPlaceGroups = await getGroups(evaNumber);
  const evaNumbers = stopPlaceGroups.STATION || [evaNumber];
  const absoluteStartTime = subMinutes(startTime, lookbehind);
  const absoluteEndTime = addMinutes(startTime, lookahead);

  const irisPromise = getAbfahrten(
    evaNumber,
    true,
    {
      lookahead,
      lookbehind,
    },
    true,
    evaNumbers,
  );
  const result = (
    await privateAxios.get('/board/v1/departures', {
      params: {
        evaNumbers,
        startTime: absoluteStartTime,
        endTime: absoluteEndTime,
      },
    })
  ).data;

  result.departures.sort(sortAbfahrt(timeByScheduled));
  result.lookbehind.sort(sortAbfahrt(timeByReal));

  const irisResult = await irisPromise;
  const irisResultByProduct: Record<string, Abfahrt> = {};
  for (const abfahrt of [
    ...irisResult.departures,
    ...irisResult.lookbehind,
    ...Object.values(irisResult.wings || {}),
  ]) {
    const key = `${abfahrt.train.number}${(
      abfahrt.departure || abfahrt.arrival
    )?.scheduledTime?.toISOString()}`;
    irisResultByProduct[key] = abfahrt;
  }

  for (const boardAbfahrt of [
    ...result.departures,
    ...result.lookbehind,
    ...Object.values(result.wings || {}),
  ]) {
    const key = `${boardAbfahrt.train.number}${(
      boardAbfahrt.departure || boardAbfahrt.arrival
    )?.scheduledTime}`;
    const irisAbfahrt = irisResultByProduct[key];
    if (irisAbfahrt) {
      boardAbfahrt.messages = irisAbfahrt.messages;
    }
  }

  return result;
}
