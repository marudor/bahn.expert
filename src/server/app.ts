/* eslint-disable unicorn/prefer-module */
import { blockMiddleware } from './blockMiddleware';
import { middlewares } from './logger';
import Axios from 'axios';
import createAdmin from './admin';
import http from 'node:http';
import Koa from 'koa';
import KoaBodyparser from 'koa-bodyparser';
import koaStatic from 'koa-static';
import path from 'node:path';
import storageMiddleware from './middleware/storageMiddleware';
import type { Context, Middleware } from 'koa';

function hotHelper(getMiddleware: () => Middleware) {
  if (process.env.NODE_ENV === 'production') {
    return getMiddleware();
  }

  return (ctx: Context, next: () => Promise<any>) => getMiddleware()(ctx, next);
}

export function createApp(): Koa {
  const app = new Koa();

  app.use(blockMiddleware);

  let apiRoutes = require('./API').default;
  let serverRender = require('./render').default;
  let seoController = require('./seo').default;
  let errorHandler = require('./errorHandler').default;
  let normalizePathMiddleware = require('./middleware/normalizePath').default;

  let devPromise = Promise.resolve();

  const distFolder = 'dist';

  if (
    process.env.NODE_ENV !== 'test' &&
    process.env.NODE_ENV !== 'production'
  ) {
    devPromise = require('./middleware/webpackDev')
      .default(app)
      .then(() => {
        app.use((ctx, next) => {
          normalizePathMiddleware =
            require('./middleware/normalizePath').default;
          errorHandler = require('./errorHandler').default;
          serverRender = require('./render').default;
          apiRoutes = require('./API').default;
          seoController = require('./seo').default;
          ctx.loadableStats = JSON.parse(
            // eslint-disable-next-line no-sync
            ctx.state.webpack.devMiddleware.outputFileSystem.readFileSync(
              path.resolve(`${distFolder}/client/loadable-stats.json`),
            ),
          );

          return next();
        });
      });
  }

  void devPromise.then(() => {
    app.use(hotHelper(() => errorHandler));
    app.use(hotHelper(() => normalizePathMiddleware()));
    app.use(storageMiddleware);
    for (const m of middlewares) app.use(m);
    app.use(KoaBodyparser());

    app.use(hotHelper(() => apiRoutes.routes()));

    app.use(
      koaStatic(
        path.resolve(
          process.env.NODE_ENV === 'production'
            ? `${distFolder}/client`
            : 'public',
        ),
        {
          maxAge: 31536000000, // 1 year
        },
      ),
    );

    app.use((ctx, next) => {
      if (
        ctx.path.endsWith('.js') ||
        ctx.path.endsWith('.css') ||
        ctx.path.endsWith('.map')
      ) {
        return;
      }
      if (ctx.url.startsWith('/api') || ctx.url.startsWith('/WRSheets')) {
        return;
      }

      return next();
    });

    app.use(hotHelper(() => seoController));

    if (process.env.NODE_ENV === 'production') {
      const loadableStats = require(
        path.resolve(`${distFolder}/client/loadable-stats.json`),
      );

      app.use((ctx, next) => {
        ctx.loadableStats = loadableStats;

        return next();
      });
    }

    app.use(hotHelper(() => serverRender));
  });

  return app;
}

export default (): http.Server => {
  const port = process.env.WEB_PORT || 9042;

  const server = http.createServer();

  Axios.defaults.baseURL = `http://localhost:${port}`;

  const app = createApp();

  // eslint-disable-next-line @typescript-eslint/no-misused-promises
  server.addListener('request', app.callback());
  server.listen(port);
  // istanbul ignore next
  if (process.env.NODE_ENV !== 'production') {
    // eslint-disable-next-line no-console
    console.log('running in DEV mode!');
  }
  createAdmin();

  return server;
};
