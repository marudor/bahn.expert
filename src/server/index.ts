globalThis.BASE_URL = `${
  process.env.NODE_ENV === 'production' && !process.env.TEST_RUN
    ? 'https://'
    : 'http://'
}${process.env.BASE_URL || 'localhost:9042'}`;
globalThis.RAW_BASE_URL = process.env.BASE_URL || 'localhost:9042';
globalThis.IMPRINT = require('./imprint');

require('./app').default();
