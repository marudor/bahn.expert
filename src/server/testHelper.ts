import { createApp } from 'server/app';
import type { Server } from 'http';

export function createTestServer(): Server {
  const app = createApp();
  const server = app.listen();

  afterAll(() => {
    server.close();
  });

  return server;
}
