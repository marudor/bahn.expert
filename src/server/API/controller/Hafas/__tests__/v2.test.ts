/* eslint-disable unicorn/prefer-module */
import { createTestServer } from 'server/testHelper';
import Nock from 'nock';
import path from 'node:path';
import request from 'supertest';

describe('Hafas V2', () => {
  const nock = Nock('https://reiseauskunft.bahn.de');
  const server = createTestServer();

  describe('/journeyDetails', () => {
    it('Parses JourneyDetails result', () => {
      nock
        .post('/bin/mgate.exe')
        .query({
          checksum: 'e9a84c4b2e75c45a5e104001fccce14f',
        })
        .replyWithFile(
          200,
          path.resolve(
            __dirname,
            '__fixtures__/journeyDetails/1|295600|0|80|10122019',
          ),
        );

      return request(server)
        .get('/api/hafas/v2/journeyDetails?jid=1|295600|0|80|10122019')
        .expect((res) => {
          expect(res.body).toMatchSnapshot();
        })
        .expect(200);
    });

    it('jid is required', () => {
      return request(server)
        .get('/api/hafas/v2/journeyDetails')
        .expect(400)
        .expect((res) => {
          expect(res.body).toMatchObject({
            fields: {
              jid: {
                message: "'jid' is required",
              },
            },
          });
        });
    });
  });
});
