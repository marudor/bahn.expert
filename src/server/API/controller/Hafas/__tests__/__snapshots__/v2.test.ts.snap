// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hafas V2 /journeyDetails Parses JourneyDetails result 1`] = `
Object {
  "firstStop": Object {
    "departure": Object {
      "platform": "5",
      "scheduledTime": "2019-12-10T21:07:00.000Z",
      "time": "2019-12-10T21:07:00.000Z",
    },
    "station": Object {
      "coordinates": Object {
        "lat": 48.948044,
        "lng": 9.137499,
      },
      "id": "8000038",
      "title": "Bietigheim-Bissingen",
    },
  },
  "jid": "1|295600|0|80|10122019",
  "lastStop": Object {
    "arrival": Object {
      "platform": "1",
      "scheduledTime": "2019-12-10T21:38:00.000Z",
      "time": "2019-12-10T21:38:00.000Z",
    },
    "station": Object {
      "coordinates": Object {
        "lat": 48.770193,
        "lng": 9.156574,
      },
      "id": "8006698",
      "title": "Stuttgart Schwabstr.",
    },
  },
  "messages": Array [
    Object {
      "code": "FB",
      "icoX": 5,
      "prio": 260,
      "txtN": "Fahrradmitnahme begrenzt möglich",
      "type": "A",
    },
    Object {
      "code": "ZB",
      "icoX": 6,
      "prio": 560,
      "txtN": "Fahrzeuggebundene Einstiegshilfe am Zuganfang",
      "type": "A",
    },
  ],
  "stops": Array [
    Object {
      "departure": Object {
        "platform": "5",
        "scheduledTime": "2019-12-10T21:07:00.000Z",
        "time": "2019-12-10T21:07:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.948044,
          "lng": 9.137499,
        },
        "id": "8000038",
        "title": "Bietigheim-Bissingen",
      },
    },
    Object {
      "arrival": Object {
        "platform": "2",
        "scheduledTime": "2019-12-10T21:10:00.000Z",
        "time": "2019-12-10T21:10:00.000Z",
      },
      "departure": Object {
        "platform": "2",
        "scheduledTime": "2019-12-10T21:10:00.000Z",
        "time": "2019-12-10T21:10:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.921679,
          "lng": 9.12575,
        },
        "id": "8005820",
        "title": "Tamm(Württ)",
      },
    },
    Object {
      "arrival": Object {
        "platform": "3",
        "scheduledTime": "2019-12-10T21:12:00.000Z",
        "time": "2019-12-10T21:12:00.000Z",
      },
      "departure": Object {
        "platform": "3",
        "scheduledTime": "2019-12-10T21:13:00.000Z",
        "time": "2019-12-10T21:13:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.907431,
          "lng": 9.147801,
        },
        "id": "8000630",
        "title": "Asperg",
      },
    },
    Object {
      "arrival": Object {
        "platform": "3",
        "scheduledTime": "2019-12-10T21:16:00.000Z",
        "time": "2019-12-10T21:16:00.000Z",
      },
      "departure": Object {
        "platform": "3",
        "scheduledTime": "2019-12-10T21:16:00.000Z",
        "time": "2019-12-10T21:16:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.892374,
          "lng": 9.185115,
        },
        "id": "8000235",
        "title": "Ludwigsburg",
      },
    },
    Object {
      "arrival": Object {
        "platform": "3",
        "scheduledTime": "2019-12-10T21:19:00.000Z",
        "time": "2019-12-10T21:19:00.000Z",
      },
      "departure": Object {
        "platform": "3",
        "scheduledTime": "2019-12-10T21:20:00.000Z",
        "time": "2019-12-10T21:20:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.862224,
          "lng": 9.179955,
        },
        "id": "8003411",
        "title": "Kornwestheim Pbf",
      },
    },
    Object {
      "arrival": Object {
        "platform": "2",
        "scheduledTime": "2019-12-10T21:23:00.000Z",
        "time": "2019-12-10T21:23:00.000Z",
      },
      "departure": Object {
        "platform": "2",
        "scheduledTime": "2019-12-10T21:24:00.000Z",
        "time": "2019-12-10T21:24:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.829647,
          "lng": 9.166669,
        },
        "id": "8005778",
        "title": "Stuttgart-Zuffenhausen",
      },
    },
    Object {
      "arrival": Object {
        "platform": "1",
        "scheduledTime": "2019-12-10T21:26:00.000Z",
        "time": "2019-12-10T21:26:00.000Z",
      },
      "departure": Object {
        "platform": "1",
        "scheduledTime": "2019-12-10T21:26:00.000Z",
        "time": "2019-12-10T21:26:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.813674,
          "lng": 9.169411,
        },
        "id": "8005770",
        "title": "Stuttgart-Feuerbach",
      },
    },
    Object {
      "arrival": Object {
        "platform": "4",
        "scheduledTime": "2019-12-10T21:28:00.000Z",
        "time": "2019-12-10T21:28:00.000Z",
      },
      "departure": Object {
        "platform": "4",
        "scheduledTime": "2019-12-10T21:29:00.000Z",
        "time": "2019-12-10T21:29:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.803399,
          "lng": 9.189331,
        },
        "id": "8005767",
        "title": "Stuttgart Nord",
      },
    },
    Object {
      "arrival": Object {
        "platform": "101",
        "scheduledTime": "2019-12-10T21:32:00.000Z",
        "time": "2019-12-10T21:32:00.000Z",
      },
      "departure": Object {
        "platform": "101",
        "scheduledTime": "2019-12-10T21:33:00.000Z",
        "time": "2019-12-10T21:33:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.783002,
          "lng": 9.179811,
        },
        "id": "8098096",
        "title": "Stuttgart Hbf (tief)",
      },
    },
    Object {
      "arrival": Object {
        "platform": "1",
        "scheduledTime": "2019-12-10T21:34:00.000Z",
        "time": "2019-12-10T21:34:00.000Z",
      },
      "departure": Object {
        "platform": "1",
        "scheduledTime": "2019-12-10T21:35:00.000Z",
        "time": "2019-12-10T21:35:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.77635,
          "lng": 9.172881,
        },
        "id": "8006700",
        "title": "Stuttgart Stadtmitte",
      },
    },
    Object {
      "arrival": Object {
        "platform": "1",
        "scheduledTime": "2019-12-10T21:36:00.000Z",
        "time": "2019-12-10T21:36:00.000Z",
      },
      "departure": Object {
        "platform": "1",
        "scheduledTime": "2019-12-10T21:36:00.000Z",
        "time": "2019-12-10T21:36:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.772647,
          "lng": 9.165743,
        },
        "id": "8006699",
        "title": "Stuttgart Feuersee",
      },
    },
    Object {
      "arrival": Object {
        "platform": "1",
        "scheduledTime": "2019-12-10T21:38:00.000Z",
        "time": "2019-12-10T21:38:00.000Z",
      },
      "station": Object {
        "coordinates": Object {
          "lat": 48.770193,
          "lng": 9.156574,
        },
        "id": "8006698",
        "title": "Stuttgart Schwabstr.",
      },
    },
  ],
  "train": Object {
    "admin": "800643",
    "line": "5",
    "name": "S 5",
    "number": "7567",
    "operator": Object {
      "icoX": 1,
      "name": "DB Regio AG S-Bahn Stuttgart",
    },
    "type": "S",
  },
}
`;
